Storage/container/one2one.h

```diff
@@ -0,0 +1,253 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef CURRENT_STORAGE_CONTAINER_ONE2ONE_H
+#define CURRENT_STORAGE_CONTAINER_ONE2ONE_H
+
+#include "common.h"
+#include "sfinae.h"
+
+#include "../base.h"
+
+#include "../../TypeSystem/optional.h"
+#include "../../Bricks/util/comparators.h"  // For `CurrentHashFunction`.
+
+namespace current {
+namespace storage {
+namespace container {
+
+template <typename T,
+          typename T_UPDATE_EVENT,
+          typename T_DELETE_EVENT,
+          template <typename...> class ROW_MAP,
+          template <typename...> class COL_MAP>
+class GenericOne2One {
+ public:
+  using T_ROW = sfinae::ENTRY_ROW_TYPE<T>;
+  using T_COL = sfinae::ENTRY_COL_TYPE<T>;
+  using T_RCPAIR = std::pair<T_ROW, T_COL>;
+  using T_WHOLE_MAP = std::unordered_map<T_RCPAIR, std::unique_ptr<T>, CurrentHashFunction<T_RCPAIR>>;
+  using T_FORWARD_MAP = ROW_MAP<T_ROW, const T*>;
+  using T_TRANSPOSED_MAP = COL_MAP<T_COL, const T*>;
+  using T_REST_BEHAVIOR = rest::behavior::Matrix;
+
+  explicit GenericOne2One(MutationJournal& journal) : journal_(journal) {}
+
+  bool Empty() const { return map_.empty(); }
+  size_t Size() const { return map_.size(); }
+
+  void Add(const T& object) {
+    const auto row = sfinae::GetRow(object);
+    const auto col = sfinae::GetCol(object);
+    const auto row_col = std::make_pair(row, col);
+    const auto it = map_.find(row_col);
+    if (it != map_.end()) {
+      const T prev_object = *(it->second);
+      journal_.LogMutation(T_UPDATE_EVENT(object),
+                           [this, row_col, prev_object]() { DoAdd(row_col, prev_object); });
+    } else {
+      const auto it_row = forward_.find(row);
+      const auto it_col = transposed_.find(col);
+      bool row_occupied = it_row != forward_.end();
+      bool col_occupied = it_col != transposed_.end();
+      if (row_occupied && col_occupied) {
+        const T prev_object1 = *(it_row->second);
+        const T prev_object2 = *(it_col->second);
```

Comment:
And another nit: in Google style guide it's generally not recommended to shorten otherwise short variable names. So, `previous_*` would always be preferred to `prev_*`.